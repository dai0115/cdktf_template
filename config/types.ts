import { VpcConfig } from "@cdktf/provider-aws/lib/vpc";

import { S3BucketConfig } from "@cdktf/provider-aws/lib/s3-bucket";
import { S3BucketAclConfig } from "@cdktf/provider-aws/lib/s3-bucket-acl";
import { S3BucketPublicAccessBlockConfig } from "@cdktf/provider-aws/lib/s3-bucket-public-access-block";
import { Wafv2IpSetConfig } from "@cdktf/provider-aws/lib/wafv2-ip-set";
import { Wafv2RuleGroupConfig } from "@cdktf/provider-aws/lib/wafv2-rule-group";
import { Wafv2WebAclConfig } from "@cdktf/provider-aws/lib/wafv2-web-acl";
import { CloudfrontDistributionConfig } from "@cdktf/provider-aws/lib/cloudfront-distribution";
import { CloudfrontOriginAccessControlConfig } from "@cdktf/provider-aws/lib/cloudfront-origin-access-control";
import { CognitoUserPoolConfig } from "@cdktf/provider-aws/lib/cognito-user-pool";
import { EcrRepositoryConfig } from "@cdktf/provider-aws/lib/ecr-repository";
import { EcsClusterConfig } from "@cdktf/provider-aws/lib/ecs-cluster";
import { EcsServiceConfig } from "@cdktf/provider-aws/lib/ecs-service";
import { EcsTaskDefinitionConfig } from "@cdktf/provider-aws/lib/ecs-task-definition";
import {
  CodedeployDeploymentGroupAutoRollbackConfiguration,
  CodedeployDeploymentGroupBlueGreenDeploymentConfig,
  CodedeployDeploymentGroupDeploymentStyle,
} from "@cdktf/provider-aws/lib/codedeploy-deployment-group";
import { IamRoleConfig } from "@cdktf/provider-aws/lib/iam-role";

export type stageType = "dev" | "staging" | "prduction";

type prefix = `application-${stageType}`;

// Stack毎に可変なプロパティを定義
export type ConfigType = {
  region: string;
  prefix: prefix;
  stage: stageType;
  network: NetworkConfig;
  db: DbConfig;
  cognito: CognitoConfigType;
  staticWebsiteHosting: StaticWebsiteHostingConfigType;
  computing: ComputingConfigType;
};

type NetworkConfig = {
  vpcConfig: VpcConfig;
};

type DbConfig = {
  dbMasterUsername: string;
  dbFamily: string;
  clusterEngine: string;
  clusterEngineVersion: string;
  instanceCount: number;
  instanceClass: string;
};

type OptionalCognitoConfig =
  | "aliasAttributes"
  | "deviceConfiguration"
  | "emailVerificationMessage"
  | "emailVerificationSubject"
  | "smsAuthenticationMessage"
  | "smsConfiguration"
  | "smsVerificationMessage"
  | "tags"
  | "userAttributeUpdateSettings"
  | "userPoolAddOns"
  | "verificationMessageTemplate";

type CognitoConfigType = Omit<CognitoUserPoolConfig, OptionalCognitoConfig>;

type StaticWebsiteHostingConfigType = {
  s3: S3ConfigType;
  bucketAcl: S3AclType;
  ipSet: IPSetType;
  ruleGroup: RuleGroupConfigType;
  wafv2WebAcl: Wafv2WebAclType;
  cloudfrontOAC: CloudfrontOriginAccessControlConfigType;
  cloudfrontDistribution: CloudfrontDistributionWithoutType;
  s3PublicAccessBlock: S3PublicAccessBlockType;
};

// SPA（App）用S3バケットS3の型定義
type OptionalS3BucketConfig = "none";
type OptionalS3BucketAclConfig = "bucket";
type S3ConfigType = Omit<S3BucketConfig, OptionalS3BucketConfig>;
type S3AclType = Omit<S3BucketAclConfig, OptionalS3BucketAclConfig>;

// IPSetの型定義
type OptionalIpSetConfig = "provider";
type IPSetType = Omit<Wafv2IpSetConfig, OptionalIpSetConfig>;

// WafRuleGroupの型定義
type OptionalRuleGroupConfig = "name";
type RuleGroupConfigType = Omit<Wafv2RuleGroupConfig, OptionalRuleGroupConfig>;

// WafAclの型定義
type OptinalWafv2WebAclConfig = "name" | "provider";
type Wafv2WebAclType = Omit<Wafv2WebAclConfig, OptinalWafv2WebAclConfig>;

// cloudfrontOACの型定義
type OptinalCloudfrontOriginAccessControlConfig = "name";
type CloudfrontOriginAccessControlConfigType = Omit<
  CloudfrontOriginAccessControlConfig,
  OptinalCloudfrontOriginAccessControlConfig
>;

// Cloudfront Distributionの型定義
type OptionalCloudfrontDistributionType = "origin" | "defaultCacheBehavior";
type CloudfrontDistributionWithoutType = Omit<
  CloudfrontDistributionConfig,
  OptionalCloudfrontDistributionType
>;

type OptinalS3BucketPublicAccessBlockConfig = "bucket";
type S3PublicAccessBlockType = Omit<
  S3BucketPublicAccessBlockConfig,
  OptinalS3BucketPublicAccessBlockConfig
>;

type ComputingConfigType = {
  ecrConfig: Omit<EcrRepositoryConfig, "name">;
  clusterConfig: Omit<EcsClusterConfig, "name">;
  taskRoleConfig: Omit<IamRoleConfig, "name" | "assumeRolePolicy">;
  executionRoleConfig: Omit<IamRoleConfig, "name" | "assumeRolePolicy">;
  taskDefinitionConfig: Omit<
    EcsTaskDefinitionConfig,
    | "name"
    | "family"
    | "taskRoleArn"
    | "executionRoleArn"
    | "containerDefinitions"
  >;
  serviceConfig: Omit<
    EcsServiceConfig,
    | "name"
    | "cluster"
    | "taskDefinition"
    | "networkConfiguration"
    | "loadBalancer"
  >;
  autoRollbackConfig: CodedeployDeploymentGroupAutoRollbackConfiguration;
  bluegreenDeploymentConfig: CodedeployDeploymentGroupBlueGreenDeploymentConfig;
  deploymentStyleConfig: CodedeployDeploymentGroupDeploymentStyle;
  deploymentConfigName: string;
};
