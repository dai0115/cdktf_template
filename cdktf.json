{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "8f6cdb9b-f601-410e-ace2-9b4a7e619262",
  "sendCrashReports": "false",
  "terraformProviders": [],
  "terraformModules": [],
  "context": {
    "excludeStackIdFromLogicalIds": "true",
    "allowSepCharsInLogicalIds": "true"
  }
}
