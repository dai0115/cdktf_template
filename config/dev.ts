import { ConfigType } from "./types";

export const devConfig: ConfigType = {
  region: "ap-northeast-1",
  stage: "dev",
  prefix: "application-dev",
  network: {
    vpcConfig: {
      cidrBlock: "10.0.0.0/16",
      enableDnsHostnames: true,
      enableDnsSupport: true,
    },
  },
};
