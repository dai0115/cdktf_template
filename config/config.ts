import { ConfigType, stageType } from "./types";
import { devConfig } from "./dev";

// TODO stagingとproductionを作る際に@ts-ignoreを削除する
const configMap: { [key in stageType]: ConfigType } = {
  dev: devConfig,
  // @ts-ignore
  staging: {},
  // @ts-ignore
  prduction: {},
};

export const getConfig = (stage: stageType): ConfigType => {
  return configMap[stage];
};
