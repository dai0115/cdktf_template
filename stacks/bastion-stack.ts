import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { TerraformStack } from "cdktf";
import { Construct } from "constructs";

import { EcrRepository } from "@cdktf/provider-aws/lib/ecr-repository";
import { EcsTaskDefinition } from "@cdktf/provider-aws/lib/ecs-task-definition";
import { EcsCluster } from "@cdktf/provider-aws/lib/ecs-cluster";
import { DataAwsIamPolicyDocument } from "@cdktf/provider-aws/lib/data-aws-iam-policy-document";
import { IamRole } from "@cdktf/provider-aws/lib/iam-role";
import { DataAwsCallerIdentity } from "@cdktf/provider-aws/lib/data-aws-caller-identity";

import { ConfigType } from "../config/types";

type BastionProps = {
  config: ConfigType;
};

export class BastionStack extends TerraformStack {
  constructor(scope: Construct, id: string, props: BastionProps) {
    super(scope, id);
    const { region, prefix } = props.config;

    new AwsProvider(this, `${prefix}-AWS`, {
      region: region,
    });

    const accountId = new DataAwsCallerIdentity(this, "account-id").accountId;

    // 踏み台のイメージを保存するためのリポジトリを作成
    const repo = new EcrRepository(this, "ecr", {
      name: `${prefix}-ecr-bastion-repo`,
      imageScanningConfiguration: { scanOnPush: true },
      encryptionConfiguration: [
        {
          encryptionType: "KMS",
        },
      ],
    });

    const taskAssumeRole = this.createAssumeRolePolicy(
      `${prefix}-task`,
      "ecs-tasks.amazonaws.com"
    );

    // タスクロールの作成
    const taskRole = new IamRole(this, `${prefix}-task-role`, {
      name: `${prefix}-task-role`,
      assumeRolePolicy: taskAssumeRole.json,
      inlinePolicy: [
        {
          name: "taskRole",
          policy: JSON.stringify({
            Version: "2012-10-17",
            Statement: [
              {
                Effect: "Allow",
                Action: [
                  "ssmmessages:CreateActivation",
                  "ssmmessages:RemoveTagsFromResource",
                  "ssmmessages:AddTagsFromResource",
                  "ssmmessages:DeleteActivation",
                  "ssmmessages:CreateControlChannel",
                  "ssmmessages:CreateDataChannel",
                  "ssmmessages:OpenControlChannel",
                  "ssmmessages:OpenDataChannel",
                ],
                Resource: "*",
              },
            ],
          }),
        },
      ],
    });
    // タスク実行ロールの作成
    const taskExecutionRole = new IamRole(
      this,
      `${prefix}-task-execution-role`,
      {
        name: `${prefix}-task-execution-role`,
        assumeRolePolicy: taskAssumeRole.json,
        inlinePolicy: [
          {
            name: "taskExecutionRole",
            policy: JSON.stringify({
              Version: "2012-10-17",
              Statement: [
                {
                  Effect: "Allow",
                  Action: [
                    "ecr:GetAuthorizationToken",
                    "ecr:BatchCheckLayerAvailability",
                    "ecr:GetDownloadUrlForLayer",
                    "ecr:BatchGetImage",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                    "secretsmanager:GetSecret",
                    "secretsmanager:GetSecretValue",
                    "secretsmanager:DescribeSecret",
                    "secretsmanager:ListSecrets",
                    "logs:*",
                  ],
                  Resource: "*",
                },
              ],
            }),
          },
        ],
      }
    );

    // ECS関連のリソースを作成
    new EcsCluster(this, `${prefix}-BastionCluster`, {
      name: `${prefix}-BastionCluster`,
      setting: [
        {
          name: "containerInsights",
          value: "enabled",
        },
      ],
    });

    new EcsTaskDefinition(this, `${prefix}-BastiontaskDefinition`, {
      family: `${prefix}-BastiontaskDefinition`,
      taskRoleArn: taskRole.arn,
      executionRoleArn: taskExecutionRole.arn,
      requiresCompatibilities: ["FARGATE"],
      cpu: "256",
      memory: "512",
      networkMode: "awsvpc",
      containerDefinitions: JSON.stringify([
        {
          name: "bastion",
          image: `${accountId}.dkr.ecr.ap-northeast-1.amazonaws.com/${repo.name}:latest`,
          essential: true,
          linuxParameters: {
            initProcessEnabled: true,
          },
          cpu: 256,
          memory: 512,
          portMappings: [
            {
              containerPort: 80,
            },
          ],
          logConfiguration: {
            logDriver: "awslogs",
            options: {
              "awslogs-region": "ap-northeast-1",
              "awslogs-stream-prefix": "ecs-bastion-task",
              "awslogs-group": "/ecs/ecs-bastion-task",
              "awslogs-create-group": "true",
            },
          },
        },
      ]),
    });
  }

  /**
   * @param prefix - リソース命名用のprefix
   * @param identifiers - アクセスするAWSサービスのidentifiersを指定
   * @returns 作成したロールをJSON形式で返却
   */
  private createAssumeRolePolicy(
    prefix: string,
    identifiers: string
  ): DataAwsIamPolicyDocument {
    const assumeRolePolicy = new DataAwsIamPolicyDocument(
      this,
      `${prefix}-assumeRolePolicy`,
      {
        statement: [
          {
            effect: "Allow",
            actions: ["sts:AssumeRole"],
            principals: [
              {
                identifiers: [identifiers],
                type: "Service",
              },
            ],
          },
        ],
      }
    );
    return assumeRolePolicy;
  }
}
