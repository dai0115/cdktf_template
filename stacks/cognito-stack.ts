import { CognitoUserPool } from "@cdktf/provider-aws/lib/cognito-user-pool";
import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { TerraformStack } from "cdktf";
import { Construct } from "constructs";
import { ConfigType } from "../config/types";

type CognitoProps = {
  config: ConfigType;
};

export class CognitoStack extends TerraformStack {
  readonly cognitoId: String;

  constructor(scope: Construct, id: string, props: CognitoProps) {
    super(scope, id);
    const { region, prefix } = props.config;
    const { cognito } = props.config;

    new AwsProvider(this, `${prefix}-AWS`, {
      region: region,
    });

    const cog = new CognitoUserPool(this, `${prefix}-Cognito`, {
      ...cognito,
    });
    this.cognitoId = cog.id;
  }
}
