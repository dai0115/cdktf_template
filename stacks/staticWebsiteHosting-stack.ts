import { CloudfrontDistribution } from "@cdktf/provider-aws/lib/cloudfront-distribution";
import { CloudfrontOriginAccessControl } from "@cdktf/provider-aws/lib/cloudfront-origin-access-control";
import { DataAwsCallerIdentity } from "@cdktf/provider-aws/lib/data-aws-caller-identity";
import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { S3Bucket } from "@cdktf/provider-aws/lib/s3-bucket";
import { S3BucketAcl } from "@cdktf/provider-aws/lib/s3-bucket-acl";
import { S3BucketPolicy } from "@cdktf/provider-aws/lib/s3-bucket-policy";
import { S3BucketPublicAccessBlock } from "@cdktf/provider-aws/lib/s3-bucket-public-access-block";
import { Wafv2IpSet } from "@cdktf/provider-aws/lib/wafv2-ip-set";
import { Wafv2WebAcl } from "@cdktf/provider-aws/lib/wafv2-web-acl";
import { TerraformStack } from "cdktf";
import { Construct } from "constructs";
import { ConfigType } from "../config/types";

type StaticWebsiteHostingProps = {
  config: ConfigType;
};

export class StaticWebsiteHostingStack extends TerraformStack {
  constructor(scope: Construct, id: string, props: StaticWebsiteHostingProps) {
    super(scope, id);
    const { prefix } = props.config;
    const {
      s3,
      bucketAcl,
      ipSet,
      wafv2WebAcl,
      cloudfrontOAC,
      cloudfrontDistribution,
      s3PublicAccessBlock,
    } = props.config.staticWebsiteHosting;

    const provider = new AwsProvider(this, `${prefix}-AWS-region-virginia`, {
      region: "us-east-1", //右記URLの制約により、リージョンをus-eas-1に変更。https://docs.aws.amazon.com/ja_jp/waf/latest/developerguide/how-aws-waf-works.html#how-aws-waf-works-resources
      alias: "virginia",
    });

    const accountId = new DataAwsCallerIdentity(this, "account-id").accountId;

    const authBucket = new S3Bucket(this, `${prefix}-s3-bucket`, {
      ...s3,
    });
    new S3BucketAcl(this, `${prefix}-s3-acl`, {
      bucket: authBucket.bucket,
      ...bucketAcl,
    });

    const caasipSet = new Wafv2IpSet(this, `${prefix}-waf-ipset`, {
      provider: provider,
      ...ipSet,
    });

    const waf = new Wafv2WebAcl(this, `${prefix}-waf-acl`, {
      ...wafv2WebAcl,
      name: `${prefix}-waf-acl`,
      provider: provider,
      rule: [
        {
          name: "AllowIpsRule",
          priority: 0,
          action: { allow: {} },
          statement: {
            ipSetReferenceStatement: { arn: caasipSet.arn },
          },
          visibilityConfig: {
            cloudwatchMetricsEnabled: true,
            metricName: "AllowedIPMetric",
            sampledRequestsEnabled: true,
          },
        },
        {
          name: "AWS-AWSManagedRulesCommonRuleSet",
          priority: 1,
          overrideAction: {
            none: {},
          },
          statement: {
            managedRuleGroupStatement: {
              name: "AWSManagedRulesCommonRuleSet",
              vendorName: "AWS",
            },
          },
          visibilityConfig: {
            cloudwatchMetricsEnabled: true,
            metricName: "AWS-AWSManagedRulesCommonRuleSet",
            sampledRequestsEnabled: true,
          },
        },
        {
          name: "AWS-AWSManagedRulesKnownBadInputsRuleSet",
          priority: 2,
          overrideAction: {
            none: {},
          },
          statement: {
            managedRuleGroupStatement: {
              name: "AWSManagedRulesKnownBadInputsRuleSet",
              vendorName: "AWS",
            },
          },
          visibilityConfig: {
            cloudwatchMetricsEnabled: true,
            metricName: "AWS-AWSManagedRulesKnownBadInputsRuleSet",
            sampledRequestsEnabled: true,
          },
        },
        {
          name: "AWS-AWSManagedRulesAdminProtectionRuleSet",
          priority: 3,
          overrideAction: {
            none: {},
          },
          statement: {
            managedRuleGroupStatement: {
              name: "AWSManagedRulesAdminProtectionRuleSet",
              vendorName: "AWS",
            },
          },
          visibilityConfig: {
            cloudwatchMetricsEnabled: true,
            metricName: "AWS-AWSManagedRulesAdminProtectionRuleSet",
            sampledRequestsEnabled: true,
          },
        },
      ],
    });

    const oac = new CloudfrontOriginAccessControl(
      this,
      `${prefix}-cloudfront-OAC`,
      {
        ...cloudfrontOAC,
        name: authBucket.id,
      }
    );

    const distribution = new CloudfrontDistribution(
      this,
      `${prefix}-cloudfront-distribution`,
      {
        webAclId: waf.arn,
        origin: [
          {
            domainName: authBucket.bucketRegionalDomainName,
            originAccessControlId: oac.id,
            originId: authBucket.bucket,
          },
        ],
        defaultCacheBehavior: {
          forwardedValues: {
            queryString: false,
            cookies: {
              forward: "none",
            },
            headers: [
              "Accept-Datetime",
              "Accept-Encoding",
              "Accept-Language",
              "Origin",
            ],
          },
          allowedMethods: ["GET", "HEAD"],
          cachedMethods: ["GET", "HEAD"],
          viewerProtocolPolicy: "allow-all",
          targetOriginId: authBucket.bucket,
        },
        ...cloudfrontDistribution,
      }
    );

    new S3BucketPublicAccessBlock(this, `${prefix}-s3-PAB`, {
      bucket: authBucket.id,
      ...s3PublicAccessBlock,
    });

    new S3BucketPolicy(this, `${prefix}-s3-policy`, {
      bucket: authBucket.bucket,
      policy: JSON.stringify({
        Version: "2012-10-17",
        Statement: [
          {
            Sid: "AllowCloudFrontServicePrincipalReadOnly",
            Effect: "Allow",
            Principal: {
              Service: "cloudfront.amazonaws.com",
            },
            Action: "s3:GetObject",
            Resource: `arn:aws:s3:::${authBucket.bucket}/*`,
            Condition: {
              StringEquals: {
                "AWS:SourceArn": `arn:aws:cloudfront::${accountId}:distribution/${distribution.id}`,
              },
            },
          },
        ],
      }),
    });
  }
}
