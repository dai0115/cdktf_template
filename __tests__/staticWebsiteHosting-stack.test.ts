import { Testing } from "cdktf";
import { devConfig } from "../config/dev";
import { StaticWebsiteHostingStack } from "../stacks/staticWebsiteHosting-stack";

describe("staticWebsiteHostingStack", () => {
  it("snapshot test for staticWebsiteHostingStack", () => {
    const app = Testing.app();
    const staticWebsiteHosting = new StaticWebsiteHostingStack(
      app,
      "staticWebsiteHostingStack",
      {
        config: devConfig,
      }
    );
    expect(Testing.synth(staticWebsiteHosting)).toMatchSnapshot();
  });
});
