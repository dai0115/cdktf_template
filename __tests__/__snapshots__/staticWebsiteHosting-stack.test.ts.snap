// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`staticWebsiteHostingStack snapshot test for staticWebsiteHostingStack 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "account-id": {
      }
    }
  },
  "provider": {
    "aws": [
      {
        "alias": "virginia",
        "region": "us-east-1"
      }
    ]
  },
  "resource": {
    "aws_cloudfront_distribution": {
      "application-dev-cloudfront-distribution": {
        "default_cache_behavior": {
          "allowed_methods": [
            "GET",
            "HEAD"
          ],
          "cached_methods": [
            "GET",
            "HEAD"
          ],
          "forwarded_values": {
            "cookies": {
              "forward": "none"
            },
            "headers": [
              "Accept-Datetime",
              "Accept-Encoding",
              "Accept-Language",
              "Origin"
            ],
            "query_string": false
          },
          "target_origin_id": "\${aws_s3_bucket.application-dev-s3-bucket.bucket}",
          "viewer_protocol_policy": "allow-all"
        },
        "default_root_object": "index.html",
        "enabled": true,
        "origin": [
          {
            "domain_name": "\${aws_s3_bucket.application-dev-s3-bucket.bucket_regional_domain_name}",
            "origin_access_control_id": "\${aws_cloudfront_origin_access_control.application-dev-cloudfront-OAC.id}",
            "origin_id": "\${aws_s3_bucket.application-dev-s3-bucket.bucket}"
          }
        ],
        "price_class": "PriceClass_200",
        "restrictions": {
          "geo_restriction": {
            "locations": [
              "JP"
            ],
            "restriction_type": "whitelist"
          }
        },
        "viewer_certificate": {
          "cloudfront_default_certificate": true
        },
        "web_acl_id": "\${aws_wafv2_web_acl.application-dev-waf-acl.arn}"
      }
    },
    "aws_cloudfront_origin_access_control": {
      "application-dev-cloudfront-OAC": {
        "name": "\${aws_s3_bucket.application-dev-s3-bucket.id}",
        "origin_access_control_origin_type": "s3",
        "signing_behavior": "always",
        "signing_protocol": "sigv4"
      }
    },
    "aws_s3_bucket": {
      "application-dev-s3-bucket": {
        "bucket": "application-dev-cdktf-template"
      }
    },
    "aws_s3_bucket_acl": {
      "application-dev-s3-acl": {
        "acl": "private",
        "bucket": "\${aws_s3_bucket.application-dev-s3-bucket.bucket}"
      }
    },
    "aws_s3_bucket_policy": {
      "application-dev-s3-policy": {
        "bucket": "\${aws_s3_bucket.application-dev-s3-bucket.bucket}",
        "policy": "{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Sid\\":\\"AllowCloudFrontServicePrincipalReadOnly\\",\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"Service\\":\\"cloudfront.amazonaws.com\\"},\\"Action\\":\\"s3:GetObject\\",\\"Resource\\":\\"arn:aws:s3:::\${aws_s3_bucket.application-dev-s3-bucket.bucket}/*\\",\\"Condition\\":{\\"StringEquals\\":{\\"AWS:SourceArn\\":\\"arn:aws:cloudfront::\${data.aws_caller_identity.account-id.account_id}:distribution/\${aws_cloudfront_distribution.application-dev-cloudfront-distribution.id}\\"}}}]}"
      }
    },
    "aws_s3_bucket_public_access_block": {
      "application-dev-s3-PAB": {
        "block_public_acls": true,
        "block_public_policy": true,
        "bucket": "\${aws_s3_bucket.application-dev-s3-bucket.id}",
        "ignore_public_acls": true,
        "restrict_public_buckets": true
      }
    },
    "aws_wafv2_ip_set": {
      "application-dev-waf-ipset": {
        "ip_address_version": "IPV4",
        "name": "ipSet",
        "provider": "aws.virginia",
        "scope": "CLOUDFRONT"
      }
    },
    "aws_wafv2_web_acl": {
      "application-dev-waf-acl": {
        "default_action": {
          "block": {
          }
        },
        "description": "Web ACL",
        "name": "application-dev-waf-acl",
        "provider": "aws.virginia",
        "rule": [
          {
            "action": {
              "allow": {
              }
            },
            "name": "AllowIpsRule",
            "priority": 0,
            "statement": {
              "ip_set_reference_statement": {
                "arn": "\${aws_wafv2_ip_set.application-dev-waf-ipset.arn}"
              }
            },
            "visibility_config": {
              "cloudwatch_metrics_enabled": true,
              "metric_name": "AllowedIPMetric",
              "sampled_requests_enabled": true
            }
          },
          {
            "name": "AWS-AWSManagedRulesCommonRuleSet",
            "override_action": {
              "none": {
              }
            },
            "priority": 1,
            "statement": {
              "managed_rule_group_statement": {
                "name": "AWSManagedRulesCommonRuleSet",
                "vendor_name": "AWS"
              }
            },
            "visibility_config": {
              "cloudwatch_metrics_enabled": true,
              "metric_name": "AWS-AWSManagedRulesCommonRuleSet",
              "sampled_requests_enabled": true
            }
          },
          {
            "name": "AWS-AWSManagedRulesKnownBadInputsRuleSet",
            "override_action": {
              "none": {
              }
            },
            "priority": 2,
            "statement": {
              "managed_rule_group_statement": {
                "name": "AWSManagedRulesKnownBadInputsRuleSet",
                "vendor_name": "AWS"
              }
            },
            "visibility_config": {
              "cloudwatch_metrics_enabled": true,
              "metric_name": "AWS-AWSManagedRulesKnownBadInputsRuleSet",
              "sampled_requests_enabled": true
            }
          },
          {
            "name": "AWS-AWSManagedRulesAdminProtectionRuleSet",
            "override_action": {
              "none": {
              }
            },
            "priority": 3,
            "statement": {
              "managed_rule_group_statement": {
                "name": "AWSManagedRulesAdminProtectionRuleSet",
                "vendor_name": "AWS"
              }
            },
            "visibility_config": {
              "cloudwatch_metrics_enabled": true,
              "metric_name": "AWS-AWSManagedRulesAdminProtectionRuleSet",
              "sampled_requests_enabled": true
            }
          }
        ],
        "scope": "CLOUDFRONT",
        "visibility_config": {
          "cloudwatch_metrics_enabled": false,
          "metric_name": "WebACLMetric",
          "sampled_requests_enabled": false
        }
      }
    }
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "4.60.0"
      }
    }
  }
}"
`;
